File: examples/Python/weighted_regression.py
"""Weighted Least Squares (WLS) regression example.

Run with:
    pip install linreg-core
    python weighted_regression.py

Demonstrates:
- Why OLS fails on heteroscedastic data (variance grows with income)
- How WLS corrects for this using precision weights (1/variance)
- Side-by-side coefficient and SE comparison: OLS vs WLS
- Sanity check: equal weights reproduces OLS exactly
"""

import linreg_core


def main():
    print("╔══════════════════════════════════════════════════════════════════════╗")
    print("║            WEIGHTED LEAST SQUARES (WLS) REGRESSION                  ║")
    print("╚══════════════════════════════════════════════════════════════════════╝")
    print()

    # Dataset: income ($k) vs. spending ($k)
    # Variance in spending grows with income — classic heteroscedasticity.
    income   = [20.0, 25.0, 30.0, 35.0, 40.0, 50.0, 60.0, 75.0, 90.0, 110.0]
    spending = [18.0, 21.0, 25.0, 27.0, 31.0, 38.0, 50.0, 55.0, 72.0, 95.0]

    # Precision weights = 1 / variance = 1 / sd²
    std_devs = [1.0, 1.2, 1.5, 2.0, 2.5, 3.5, 5.0, 7.0, 9.0, 12.0]
    weights  = [1.0 / (s * s) for s in std_devs]

    print("Dataset: 10 households — income vs. spending")
    print("Heteroscedasticity: spending variance grows with income.")
    print()

    # ── 1. Data and Weights ───────────────────────────────────────────────────
    print("━━━ 1. Data and Precision Weights (w = 1/variance) ━━━━━━━━━━━━━━━━━")
    print(f"  {'Income':>8}  {'Spending':>10}  {'Std Dev':>8}  {'Weight':>10}")
    print(f"  {'─'*42}")
    for i in range(10):
        print(f"  {income[i]:>8.1f}  {spending[i]:>10.1f}  "
              f"{std_devs[i]:>8.1f}  {weights[i]:>10.4f}")
    print()
    print("  Low-income obs -> high weight (reliable).")
    print("  High-income obs -> low weight (noisy).")
    print()

    # ── 2. OLS (ignores heteroscedasticity) ───────────────────────────────────
    print("━━━ 2. OLS Regression (ignores heteroscedasticity) ━━━━━━━━━━━━━━━━━")
    names = ["Intercept", "Income"]
    ols = linreg_core.ols_regression(spending, [income], names)
    print(f"  Intercept:  {ols.coefficients[0]:>8.4f}  (SE: {ols.standard_errors[0]:.4f})")
    print(f"  Income:     {ols.coefficients[1]:>8.4f}  (SE: {ols.standard_errors[1]:.4f})")
    print(f"  R²:         {ols.r_squared:.4f}")
    print(f"  F-stat:     {ols.f_statistic:.4f}  (p = {ols.f_p_value:.6f})")
    print(f"  MSE:        {ols.mse:.4f}")
    print()
    print("  Problem: OLS gives equal weight to all observations.")
    print("  Noisy high-income points pull the line and inflate standard errors.")
    print()

    # ── 3. WLS (precision-weighted) ───────────────────────────────────────────
    print("━━━ 3. WLS Regression (precision-weighted, w = 1/variance) ━━━━━━━━━")
    wls = linreg_core.wls_regression(spending, [income], weights)
    print(f"  Intercept:   {wls.coefficients[0]:>8.4f}  (SE: {wls.standard_errors[0]:.4f})")
    print(f"  Income:      {wls.coefficients[1]:>8.4f}  (SE: {wls.standard_errors[1]:.4f})")
    print(f"  R²:          {wls.r_squared:.4f}")
    print(f"  F-stat:      {wls.f_statistic:.4f}  (p = {wls.f_p_value:.6f})")
    print(f"  Residual SE: {wls.residual_std_error:.4f}")
    print()
    print("  Fix: WLS down-weights noisy high-income observations.")
    print("  The fit is driven by reliable low-income points, tightening SEs.")
    print()

    # ── 4. Side-by-side Comparison ────────────────────────────────────────────
    print("━━━ 4. OLS vs WLS Comparison ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━")
    print(f"  {'':20}  {'OLS':>12}  {'WLS':>12}")
    print(f"  {'─'*48}")
    print(f"  {'Intercept':<20}  {ols.coefficients[0]:>12.4f}  {wls.coefficients[0]:>12.4f}")
    print(f"  {'Intercept SE':<20}  {ols.standard_errors[0]:>12.4f}  {wls.standard_errors[0]:>12.4f}")
    print(f"  {'Income slope':<20}  {ols.coefficients[1]:>12.4f}  {wls.coefficients[1]:>12.4f}")
    print(f"  {'Income slope SE':<20}  {ols.standard_errors[1]:>12.4f}  {wls.standard_errors[1]:>12.4f}")
    print(f"  {'R²':<20}  {ols.r_squared:>12.4f}  {wls.r_squared:>12.4f}")
    print(f"  {'MSE':<20}  {ols.mse:>12.4f}  {wls.residual_std_error**2:>12.4f}")
    print()

    # ── 5. Fitted Values Comparison ───────────────────────────────────────────
    print("━━━ 5. Fitted Values vs Actual ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━")
    print(f"  {'Income':>8}  {'Actual':>10}  {'OLS fit':>10}  {'WLS fit':>10}  {'Weight':>10}")
    print(f"  {'─'*54}")
    for i in range(10):
        ols_fit = ols.coefficients[0] + ols.coefficients[1] * income[i]
        print(f"  {income[i]:>8.1f}  {spending[i]:>10.2f}  "
              f"{ols_fit:>10.2f}  {wls.fitted_values[i]:>10.2f}  {weights[i]:>10.4f}")
    print()

    # ── 6. Sanity Check: equal weights → OLS ─────────────────────────────────
    print("━━━ 6. Sanity Check: Equal Weights Reproduces OLS ━━━━━━━━━━━━━━━━━━")
    wls_eq = linreg_core.wls_regression(spending, [income], [1.0] * 10)
    max_diff = max(
        abs(ols.coefficients[0] - wls_eq.coefficients[0]),
        abs(ols.coefficients[1] - wls_eq.coefficients[1]),
    )
    print(f"  OLS intercept:         {ols.coefficients[0]:.6f}")
    print(f"  WLS (equal) intercept: {wls_eq.coefficients[0]:.6f}")
    print(f"  OLS slope:             {ols.coefficients[1]:.6f}")
    print(f"  WLS (equal) slope:     {wls_eq.coefficients[1]:.6f}")
    print(f"  Max coefficient diff:  {max_diff:.2e}")
    print(f"  Matches OLS: {max_diff < 1e-8}")


if __name__ == "__main__":
    main()

File: examples/Rust/weighted_regression.rs
//! Weighted Least Squares (WLS) regression example.
//!
//! Run with:
//!   cargo run --example weighted_regression --no-default-features
//!
//! Demonstrates:
//! - Why OLS fails on heteroscedastic data (variance grows with the predictor)
//! - How WLS corrects for this by down-weighting high-variance observations
//! - Using precision weights (w = 1/variance) as the theoretically correct choice
//! - Comparing OLS vs WLS coefficients and standard errors side-by-side

use linreg_core::core::ols_regression;
use linreg_core::weighted_regression::wls_regression;

fn main() {
    // ── Dataset: income ($k) vs. spending ($k) ─────────────────────────────────
    // Variance in spending grows with income — classic heteroscedasticity.
    // Low-income households spend predictably; high-income households vary widely.
    let income  = vec![ 20.0,  25.0,  30.0,  35.0,  40.0,  50.0,  60.0,  75.0,  90.0, 110.0];
    let spending = vec![ 18.0,  21.0,  25.0,  27.0,  31.0,  38.0,  50.0,  55.0,  72.0,  95.0];

    // Approximate standard deviations: higher income -> higher variance.
    // weights = 1 / variance = 1 / sd²  (precision weighting)
    let std_devs = vec![1.0, 1.2, 1.5, 2.0, 2.5, 3.5, 5.0, 7.0, 9.0, 12.0];
    let weights: Vec<f64> = std_devs.iter().map(|s| 1.0 / (s * s)).collect();

    println!("╔══════════════════════════════════════════════════════════════════════╗");
    println!("║            WEIGHTED LEAST SQUARES (WLS) REGRESSION                  ║");
    println!("╚══════════════════════════════════════════════════════════════════════╝");
    println!();
    println!("Dataset: income vs. spending (10 households)");
    println!("Heteroscedasticity: spending variance grows with income.");
    println!();

    // ── 1. Show the data and weights ───────────────────────────────────────────
    println!("━━━ 1. Data and Precision Weights ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━");
    println!("  {:>8}  {:>10}  {:>8}  {:>8}", "Income", "Spending", "Std Dev", "Weight");
    println!("  {}", "─".repeat(42));
    for i in 0..10 {
        println!(
            "  {:>8.1}  {:>10.1}  {:>8.1}  {:>8.4}",
            income[i], spending[i], std_devs[i], weights[i]
        );
    }
    println!();
    println!("  Weights = 1/variance. Low-income obs get high weight (more reliable).");
    println!("  High-income obs get low weight (noisy, less influential).");
    println!();

    // ── 2. OLS fit (ignores heteroscedasticity) ────────────────────────────────
    println!("━━━ 2. OLS Regression (ignores heteroscedasticity) ━━━━━━━━━━━━━━━━━━");
    let names = vec!["Intercept".to_string(), "Income".to_string()];
    let ols = ols_regression(&spending, &[income.clone()], &names)
        .expect("OLS failed");

    println!("  Intercept:  {:.4}  (SE: {:.4})", ols.coefficients[0], ols.std_errors[0]);
    println!("  Income:     {:.4}  (SE: {:.4})", ols.coefficients[1], ols.std_errors[1]);
    println!("  R²:         {:.4}", ols.r_squared);
    println!("  F-stat:     {:.4}  (p = {:.6})", ols.f_statistic, ols.f_p_value);
    println!("  MSE:        {:.4}", ols.mse);
    println!();
    println!("  Problem: OLS gives equal weight to all observations. The noisy");
    println!("  high-income points pull the line and inflate standard errors.");
    println!();

    // ── 3. WLS fit (precision-weighted) ───────────────────────────────────────
    println!("━━━ 3. WLS Regression (precision-weighted, 1/variance) ━━━━━━━━━━━━━━");
    let wls = wls_regression(&spending, &[income.clone()], &weights)
        .expect("WLS failed");

    println!("  Intercept:  {:.4}  (SE: {:.4})", wls.coefficients[0], wls.standard_errors[0]);
    println!("  Income:     {:.4}  (SE: {:.4})", wls.coefficients[1], wls.standard_errors[1]);
    println!("  R²:         {:.4}", wls.r_squared);
    println!("  F-stat:     {:.4}  (p = {:.6})", wls.f_statistic, wls.f_p_value);
    println!("  Residual SE:{:.4}", wls.residual_std_error);
    println!();
    println!("  Fix: WLS down-weights noisy high-income observations. The fit is");
    println!("  driven by the reliable low-income points, tightening standard errors.");
    println!();

    // ── 4. Side-by-side comparison ────────────────────────────────────────────
    println!("━━━ 4. OLS vs WLS Comparison ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━");
    println!("  {:20}  {:>12}  {:>12}", "", "OLS", "WLS");
    println!("  {}", "─".repeat(48));
    println!("  {:20}  {:>12.4}  {:>12.4}", "Intercept",      ols.coefficients[0], wls.coefficients[0]);
    println!("  {:20}  {:>12.4}  {:>12.4}", "Intercept SE",   ols.std_errors[0],   wls.standard_errors[0]);
    println!("  {:20}  {:>12.4}  {:>12.4}", "Income slope",   ols.coefficients[1], wls.coefficients[1]);
    println!("  {:20}  {:>12.4}  {:>12.4}", "Income slope SE",ols.std_errors[1],   wls.standard_errors[1]);
    println!("  {:20}  {:>12.4}  {:>12.4}", "R²",             ols.r_squared,       wls.r_squared);
    println!("  {:20}  {:>12.4}  {:>12.4}", "MSE",            ols.mse,             wls.residual_std_error.powi(2));
    println!();

    // ── 5. Fitted values comparison ───────────────────────────────────────────
    println!("━━━ 5. Fitted Values vs Actual ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━");
    println!("  {:>8}  {:>10}  {:>10}  {:>10}  {:>10}",
             "Income", "Actual", "OLS fit", "WLS fit", "Weight");
    println!("  {}", "─".repeat(54));
    for i in 0..10 {
        let ols_fit = ols.coefficients[0] + ols.coefficients[1] * income[i];
        println!(
            "  {:>8.1}  {:>10.2}  {:>10.2}  {:>10.2}  {:>10.4}",
            income[i], spending[i], ols_fit, wls.fitted_values[i], weights[i]
        );
    }
    println!();

    // ── 6. Equal weights = OLS (sanity check) ────────────────────────────────
    println!("━━━ 6. Sanity Check: Equal Weights Reproduces OLS ━━━━━━━━━━━━━━━━━━━");
    let equal_weights = vec![1.0; 10];
    let wls_equal = wls_regression(&spending, &[income.clone()], &equal_weights)
        .expect("WLS equal weights failed");
    println!("  OLS intercept:        {:.6}", ols.coefficients[0]);
    println!("  WLS (equal) intercept:{:.6}", wls_equal.coefficients[0]);
    println!("  OLS slope:            {:.6}", ols.coefficients[1]);
    println!("  WLS (equal) slope:    {:.6}", wls_equal.coefficients[1]);
    let max_diff = (ols.coefficients[0] - wls_equal.coefficients[0]).abs()
        .max((ols.coefficients[1] - wls_equal.coefficients[1]).abs());
    println!("  Max coefficient difference: {:.2e}", max_diff);
    println!("  WLS with equal weights matches OLS: {}", max_diff < 1e-8);
}
